Sprint 2/step 1

pvl1001 opened the pull request.

Commit:
Использовать Context в конструкторе бургеров

Commit:
Подсчитывать стоимость бургера с использованием useReducer

Commit:
Оформить заказ

При нажатии на кнопку делать запрос к бэкенду. Полученный номер заказа сохранять в контекст

gennady-bars commented on src/components/BurgerConstructor/ConstructorOrder.js:
```diff
@@ -3,30 +3,59 @@ import { Button, CurrencyIcon } from "@ya.praktikum/react-developer-burger-ui-co
+            headers: { 'Content-Type': 'application/json' },
+            body: JSON.stringify( { ingredients: ingredientsId } )
+         } )
+         const { order, success } = await response.json()
```
Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.


```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

gennady-bars replied:
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

У Вас повторяется код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:

```js
function checkResponse(res) {
        // тут проверка ответа
    }

```

Ее нужно поместить в папку `utils` и импортировать туда, где нужна проверка на `ok`.

и теперь можно удалить кучу ненужно повторного кода и заменить его этим:

```js
 .then(checkResponse)

```

Обратите внимание, что передается только **ссылка** на функцию. Не нужно ее вызывать. Она сама вызовется, так как в `then` нужно передавать именно **функцию**, а не **вызов** функции.

-----------------------------------------

**Можно лучше**

Можно сделать **универсальную функцию запроса с проверкой ответа**, чтобы не дублировать эту проверку в каждом запросе:

```js

  function request(url, options) {
    // принимает два аргумента: урл и объект опций, как и `fetch`
    return fetch(url, options).then(checkResponse)
  }

```

И теперь просто нужно заменить все `fetch` на `request` и убрать дублирование проверки на `ok`. Все остальное будет без изменений. Код станет чище

gennady-bars commented on src/components/App.tsx:
```diff
@@ -36,8 +37,10 @@ function App() {
             <div className="main-row">
-               <BurgerIngredients data={ingredients}/>
-               <BurgerConstructor data={ingredients}/>
+               <IngredientsContext.Provider value={ingredients}>
```
Отличный общий контекст сделали

gennady-bars commented on src/components/BurgerConstructor/BurgerConstructor.js:
```diff
@@ -1,46 +1,65 @@
+import { IngredientsContext } from "../../context/burgerContext";
 
 
 const WithModalConstructorItem = withModal( ConstructorItem )
```
* У Вас сейчас висят 32 обработчика `Escape` одновременно. Скрин https://disk.yandex.ru/i/MI-RRDh4v-timg . Не нужно было делать `withModal.js`. Нужно было сделать простой вариант: внутри `Modal` поместить навешивание обработчика `Escape`

```jsx

isOpen & (
    <Modal
      title={'Детали ингредиента'}
      onClose={handleClose}
    >
      <IngredientDetails currentIngredient={currentIngredient} /> 
    </Modal>
)
```


при появлении попапа навешивается обработчик, а при размонтировании он будет удален

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* Отличный общий контекст сделали


но есть некоторые недочеты: 

* Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.
* У Вас будет повторяться код **проверки ответа** во всех запросах. Чтобы этого избежать, нужно сделать отдельную функцию `checkResponse`, которая будет проверять ответ от сервера:
* Картинки очень маленькие в ингредиентах https://disk.yandex.ru/i/elEy_sk3l51PFw . Нужно по макету сделать (есть 3 размера картинки в объекте ингредиента)
* В верхней булке нужно добавлять текст `верх`, а в нижней - `низ`. Посмотрите в фигме, пожалуйста
* При клике внутри бургера не нужно показывать попап ингредиента. Это только для левой части экрана, где список всех ингредиентов 
* У Вас сейчас висят 32 обработчика `Escape` одновременно. Скрин https://disk.yandex.ru/i/MI-RRDh4v-timg . Не нужно было делать `withModal.js`. Нужно было сделать простой вариант: внутри `Modal` поместить навешивание обработчика `Escape`
* Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили.


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 


Удачного рефакторинга кода.

Commit:
fix: проверка res.ok

Commit:
fix Modal:

- заменить HOC на hook
- перенести обработчик keydownHandler в Modal
- убрать открытие модального окна с карточек в конструкторе.

Commit:
fix: Картинки очень маленькие в ингредиентах

Commit:
fix: В верхней булке нужно добавлять текст верх, а в нижней - низ

Commit:
fix: убрать инлайн-стили

Commit:
fix: Картинки очень маленькие в ингредиентах

Commit:
fix: Картинки очень маленькие в ингредиентах

gennady-bars commented on src/components/App.tsx:
```diff
@@ -4,6 +4,8 @@ import BurgerConstructor from "./BurgerConstructor/BurgerConstructor";
 import {NORMA_API} from "../utils/burger-api";
+import {IngredientsContext} from "../context/burgerContext";
+import {request} from "../utils/request";
 
```
При кликах на табы `Булки, Соусы, Начинки` можно сделать логику **`плавного`** скролла до секции. Это можно сделать с помощью хука `useRef` и методом ` scrollIntoView`. Нужно установить для 3х секций 3 `useRef` и следить за изменением табов при кликах на них через функцию клика. Как меняется таб, то запускается `scrollIntoView ` со значением `{behavior: "smooth"}`
Подробнее про `useRef` тут https://ru.reactjs.org/docs/hooks-reference.html#useref
Подробнее про ` scrollIntoView` тут https://developer.mozilla.org/ru/docs/Web/API/Element/scrollIntoView

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

Commit:
upd: плавный скролл при клике табов
